docs/adr/0003-delivery-channels-templating.md

```diff
@@ -0,0 +1,92 @@
+
+## Decision
+
+A notification having necessary information needs to be translated into a certain communication object (SMS, Email, etc) before it can be sent via the desired delivery channel (SendGrid, Twillio, Airship).
+
+An example of this would be a SMS, which needs the essential information to be available for it to be converted into an SMS object before being sent to the end user.
+
+```json
+
+    public class SMS {
+        public string ToPhoneNumber { get; set; }
+        public string Body { get; set; }
+    }
+
+```
+
+Each field in the communication object will needs it's own template (property) for the associated notification type. A template for the `ToPhoneNumber` property in the above example would look something like:
+
+```
+{{ notification.propertyBag.ToPhoneNumber }}
+```
+
+### Creation of templates, how does someone create templates and in what language
+
+A template will be created based on the notification type, the communication type, and the communication object's property key.
+
+A few examples of this would be:
+
+```
+(Notification type) -> (Communication type) -> (Communication type property)
+
+New Lead            -> SMS                  -> ToPhoneNumber
+New Lead            -> SMS                  -> FromPhoneNumber
+New Lead            -> SMS                  -> Body
+
+New Lead            -> Email                -> ToEmail
+New Lead            -> Email                -> Subject
+New Lead            -> Email                -> Body
+```
+
+Every communication type will have a similar communication object, ie, an SMS will use an SMSobject that has certain set fields. Some of those fields would be required without which a notification would be impossible to be sent. Eg. A SMSObject should have `ToPhoneNumber` or EmailObject should have `ToEmail` and without these properties being populated a notification could not be sent. Furthermore, the `Body` for these 2 communication object would be equally important as there would be no point in sending an empty notification.
+
+Creation of templates can be totally flexible and dependent on the organisation which will integrate with `Marain.UserNotifications` thus every tenant will have their own set of templates for SMS and Email etc. Default templates will be provided at the start for the communication types (SMS, email, web push notifications) at start but can be modified and updated which will leave space for the organisation to configure as per their need.
+
+### Storage of templates
+
+Templates need to be stored in such a way where it can be accessed and modified easily on the fly. Thus to have this flexibilty templates will be stored in azure table storage. The template for individual properties (`ToPhoneNumber`, `FromPhoneNumber`, `ToEmail`) would have the following column headers:
+
+- Notification Type
+- Communication Type
+- Property Name (This will be the property name for the the communication object)
+- Template
```

Comment:
It's not clear what the source data for these expressions look like—you've given two mutually contradictory examples for how the SMS `ToPhoneNumber` template would work:

`{{ notification.propertyBag.ToPhoneNumber }}`

and

`{{toPhoneNumber}}`

Reply:
I have update the ADR.